Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,700 @@ void LoadSettings()
+			AddLog("ERROR MobileDock failed to open the player data file!\n");
+		}
 	}
-
 }
 
-// Temporary storage for client data to be handled in LaunchPosHook.
-CLIENT_DATA undockingShip;
+wstring GetLastBaseName(uint client)
+{
+	const auto& dockedInfo = idToDockedInfoMap[client];
+	const auto& baseInfo = Universe::get_base(dockedInfo->lastDockedSolar);
+	const auto& sysInfo = Universe::get_system(baseInfo->iSystemID);
+	wstring baseName;
+	wstring& sysName = HkGetWStringFromIDS(sysInfo->strid_name);
+	// if last base is a proxy base, it's a POB, call to base plugin for its name
+	if (((string)baseInfo->cNickname).find("_proxy_base") != string::npos)
+	{
+		LAST_PLAYER_BASE_NAME_STRUCT pobName;
+		pobName.clientID = client;
+		Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_BASE_LAST_DOCKED, &pobName);
+		baseName = L"Player base " + pobName.lastBaseName;
+	}
+	else
+	{
+		baseName = HkGetWStringFromIDS(baseInfo->iBaseIDS);
+	}
+	return baseName + L", " + sysName;
+}
 
-void __stdcall PlayerLaunch(unsigned int iShip, unsigned int client)
+JettisonResult RemoveShipFromLists(const wstring& dockedShipName, bool forcedLaunch)
 {
+	if (!nameToDockedInfoMap.count(dockedShipName))
+	{
+		return InvalidName;
+	}
+	uint dockedClientID = HkGetClientIdFromCharname(dockedShipName);
 
-	returncode = DEFAULT_RETURNCODE;
+	if (dockedClientID != -1)
+	{
+		uint ship;
+		pub::Player::GetShip(dockedClientID, ship);
+		if (ship)
+		{
+			return ShipInSpace;
+		}
+		//Force launch the ship into space if it's on the carrier.
+		if (forcedLaunch)
+		{
+			uint carrierId = HkGetClientIdFromCharname(idToDockedInfoMap[dockedClientID]->carrierName);
+			jettisonedShipsQueue[dockedClientID] = carrierId;
+			ForceLaunch(dockedClientID);
 
-	uint carrier_client = HkGetClientIdFromCharname(mobiledockClients[client].wscDockedWithCharname);
+			wstring newBaseInfo = GetLastBaseName(dockedClientID);
+			PrintUserCmdText(dockedClientID, L"You've been forcefully jettisoned by the carrier.");
+			PrintUserCmdText(dockedClientID, L"Current home base: %ls", newBaseInfo.c_str());
+		}
+		Players[dockedClientID].iLastBaseID = idToDockedInfoMap[dockedClientID]->lastDockedSolar;
+		idToDockedInfoMap.erase(dockedClientID);
+	}
+	else
+	{
+		//player offline, edit their character file to put them on last docked solar
+		CAccount* acc = HkGetAccountByCharname(dockedShipName);
+		if (acc)
+		{
+			MoveOfflineShipToLastDockedSolar(dockedShipName);
+		}
+	}
 
-	wstring clientName = (const wchar_t*)Players.GetActiveCharacterName(client);
+	wstring& carrierName = nameToDockedInfoMap[dockedShipName].carrierName;
+
+	//remove the docked ship entry from carrier's data
+	if (nameToCarrierInfoMap.count(carrierName))
+	{
+		auto& dockedList = nameToCarrierInfoMap[carrierName].dockedShipList;
+		for (auto& iter = dockedList.begin(); iter != dockedList.end(); iter++)
+		{
+			if (*iter == dockedShipName)
+			{
+				std::lock_guard<std::mutex> saveLock(saveMutex);

```

Comment:
unused?

Reply:
These are threading locks. They are scoped so they call mutex lock and then when they go out of scope they call unlock. Removing these would cause the server to crash with no indication of why.